axelor-account/src/main/java/com/axelor/apps/account/service/moveline/MoveLineServiceImpl.java

```diff
@@ -435,4 +436,40 @@ protected List<MoveLine> getMoveLinesFromIds(List<Integer> moveLineIds) {
     }
     return moveLineList;
   }
+
+  @Override
+  public void computeNewCurrencyRateOnMoveLineList(
+      Move move, BigDecimal newCurrencyRate, LocalDate dueDate) throws AxelorException {
+    for (MoveLine moveLine : move.getMoveLineList()) {
+      BigDecimal currencyAmount = BigDecimal.ZERO;
+
+      if (BigDecimal.ZERO.compareTo(moveLine.getDebit().add(moveLine.getCredit())) == 0) {
+        if (moveLine.getAccount() != null) {
+          switch (moveLine.getAccount().getCommonPosition()) {
+            case 1:
+              moveLine.setCredit(moveLine.getCurrencyAmount().multiply(newCurrencyRate));
+              break;
+            case 2:
+              moveLine.setDebit(moveLine.getCurrencyAmount().multiply(newCurrencyRate));
+              break;
+            default:
+              break;
+          }
+        }
+      }
+      currencyAmount = moveLine.getDebit().add(moveLine.getCredit());
+      currencyAmount =
+          currencyAmount.divide(
+              newCurrencyRate, AppBaseService.DEFAULT_NB_DECIMAL_DIGITS, RoundingMode.HALF_UP);
+
+      moveLine.setCurrencyAmount(currencyAmount);
+      moveLine.setCurrencyRate(newCurrencyRate);
+
+      moveLine.clearInvoiceTermList();
+      Beans.get(MoveLineInvoiceTermService.class)
+          .generateDefaultInvoiceTerm(moveLine, dueDate, false);
+
+      Beans.get(MoveLineService.class).computeFinancialDiscount(moveLine);
```

Comment:
Please use constructor injection in services methods. If this create issues with service dependencies, please create a new service

Reply:
Resolved